src/cpu/pred/bi_mode.cc

```diff
@@ -73,25 +85,37 @@ BiModeBP::BiModeBP(const BiModeBPParams &params)
  * chooses the taken array and the taken array predicts taken.
  */
 void
-BiModeBP::uncondBranch(ThreadID tid, Addr pc, void * &bpHistory)
+BiModeBP::uncondBranch(ThreadID tid, Addr pc, void * &bp_history)
 {
     BPHistory *history = new BPHistory;
     history->globalHistoryReg = globalHistoryReg[tid];
     history->takenUsed = true;
     history->takenPred = true;
     history->notTakenPred = true;
     history->finalPred = true;
-    bpHistory = static_cast<void*>(history);
-    updateGlobalHistReg(tid, true);
+    bp_history = static_cast<void*>(history);
 }
 
 void
-BiModeBP::squash(ThreadID tid, void *bpHistory)
+BiModeBP::updateHistories(ThreadID tid, Addr pc, bool uncond,
+                         bool taken, Addr target, void * &bp_history)
 {
-    BPHistory *history = static_cast<BPHistory*>(bpHistory);
+    assert(uncond || bp_history);
+    if (uncond) {
+        uncondBranch(tid, pc, bp_history);
+    }
+    updateGlobalHistReg(tid, taken);
+}
+
+
+void
+BiModeBP::squash(ThreadID tid, void * &bp_history)
+{
+    BPHistory *history = static_cast<BPHistory*>(bp_history);
     globalHistoryReg[tid] = history->globalHistoryReg;
 
     delete history;
+    bp_history = nullptr;
```

Comment:
The reason I added in all predictors the explicit nulling of the `bp_history` is I added a sanity check in the deconstructor of the `PredictorHistory`.  It fails when the history is not null and is intendet to prevent memory leaks

Reply:
I figured it was something along those lines. That makes sense.